File: 01-游戏开发/02-渲染/图形管线/计算着色器执行模型/python/main.py
# -*- coding: utf-8 -*-
"""计算着色器(Compute Shader)执行模型软件模拟。

依据 Microsoft Learn "Compute Shader Overview" / "numthreads" / "Dispatch" 归纳:
  - 线程组织:Dispatch(gx,gy,gz) 启动 gx*gy*gz 个线程组;
    [numthreads(x,y,z)] 声明每组线程数,组内标识 SV_GroupThreadID,
    全局标识 SV_DispatchThreadID = GroupID*numthreads + GroupThreadID,
    组内线性索引 SV_GroupIndex = z*X*Y + y*X + x;
  - cs_5_0 限制:每组最多 1024 线程,Z 维最多 64,Dispatch 每维最多 65535 组;
    cs_4_x 更紧(768/16KB 共享内存/无原子指令);
  - groupshared 内存 + GroupMemoryBarrierWithGroupSync 实现
    线程组内协作的树形归约(parallel reduction);
  - 原子指令 InterlockedAdd 保证多线程累加不丢更新。
"""
import random

# ---------- 线程标识 ----------

class ThreadID:
    """一个线程的四个系统值标识(MS Learn 文档口径)。"""

    def __init__(self, numthreads, group_id, group_thread_id):
        self.group_id = group_id                    # SV_GroupID
        self.group_thread_id = group_thread_id      # SV_GroupThreadID
        self.dispatch_thread_id = tuple(            # SV_DispatchThreadID
            self.group_id[i] * numthreads[i] + self.group_thread_id[i] for i in range(3))
        x, y, _ = numthreads
        gx, gy, gz = group_thread_id
        self.group_index = gz * x * y + gy * x + gx  # SV_GroupIndex


def enumerate_threads(dispatch, numthreads):
    """枚举一次 Dispatch 的全部线程,验证 4 个标识的自洽性。"""
    out = []
    for gz in range(dispatch[2]):
        for gy in range(dispatch[1]):
            for gx in range(dispatch[0]):
                for tz in range(numthreads[2]):
                    for ty in range(numthreads[1]):
                        for tx in range(numthreads[0]):
                            out.append(ThreadID(numthreads, (gx, gy, gz), (tx, ty, tz)))
    return out


# ---------- cs_5_0 / cs_4_x 硬件限制(MS Learn 原文数字) ----------

CS5 = {
    "max_threads_per_group": 1024,
    "max_z": 64,
    "max_dispatch_dim": 65535,
    "groupshared_bytes": 32 * 1024,   # 32 KB
    "atomics": True,
}
CS4X = {
    "max_threads_per_group": 768,
    "max_z": 1,
    "max_dispatch_dim": 65535,
    "groupshared_bytes": 16 * 1024,   # 16 KB
    "atomics": False,
}


def validate_numthreads(numthreads, profile):
    x, y, z = numthreads
    if x * y * z > profile["max_threads_per_group"]:
        return False, "threads per group exceed %d" % profile["max_threads_per_group"]
    if z > profile["max_z"]:
        return False, "Z dimension %d exceed %d" % (z, profile["max_z"])
    return True, ""


def validate_dispatch(dispatch, profile):
    for d in dispatch:
        if d > profile["max_dispatch_dim"]:
            return False, "dispatch dimension exceed %d" % profile["max_dispatch_dim"]
    return True, ""


# ---------- groupshared + 屏障:树形归约 ----------

class ThreadGroupSim:
    """模拟一个线程组的 groupshared 内存与 GroupMemoryBarrierWithGroupSync。

    屏障语义:全部线程到齐前谁也不能越过 → 等价于"分阶段执行",
    每阶段内线程顺序任意、互不可见对方的本阶段写入。
    """

    def __init__(self, numthreads):
        self.numthreads = numthreads
        self.shared = {}          # groupshared 内存
        self.phase = 0           # 当前阶段号(每个 barrier 递增)

    def barrier(self):
        self.phase += 1

    def run_reduction(self, values):
        """树形归约:每线程写入自己的值,逐级折半,线程 0 写出组结果。"""
        x, y, z = self.numthreads
        n = x * y * z
        assert len(values) == n
        # 阶段 0:每线程写 groupshared[i] = values[i]
        for i in range(n):
            self.shared[i] = values[i]
        self.barrier()
        stride = n // 2
        while stride > 0:
            # 本阶段:线程 i(在其值有效时)做 shared[i] += shared[i+stride]
            for i in range(stride):
                self.shared[i] += self.shared[i + stride]
            self.barrier()
            stride //= 2
        return self.shared[0]


# ---------- 原子指令:InterlockedAdd ----------

def interlocked_add_example(values, use_atomic):
    """多线程对同一计数器累加:非原子会丢更新(模拟特定交错),原子不丢。

    用固定交错模拟:线程按两批交错执行读-改-写。
    """
    if use_atomic:
        counter = 0
        for v in values:  # 原子:读改写不可分割
            counter += v
        return counter
    # 非原子:所有线程先读到旧值 0,再各自写回自己的累加 → 只留最后写的
    # (这正是 GPU 上大量线程同时 RMW 同一地址的灾难形态)
    reads = [0] * len(values)
    writes = [r + v for r, v in zip(reads, values)]
    return max(writes) if writes else 0  # 丢失了除"最后写入"外的全部更新


# ---------- 自检 ----------

def main():
    # 1) MS Learn 文档的标准例子:Dispatch(5,3,2) + numthreads(10,8,3)
    nt = (10, 8, 3)
    tid = ThreadID(nt, (2, 1, 0), (7, 5, 0))
    assert tid.dispatch_thread_id == (27, 13, 0), tid.dispatch_thread_id
    assert tid.group_index == 57, tid.group_index
    # 全量枚举自检:DispatchThreadID 全局唯一且无洞
    threads = enumerate_threads((5, 3, 2), nt)
    assert len(threads) == 5 * 3 * 2 * 10 * 8 * 3
    seen = set()
    for t in threads:
        seen.add(t.dispatch_thread_id)
        # GroupIndex 与 GroupThreadID 一一对应(同一组内)
        x, y, z = t.group_thread_id
        assert t.group_index == z * nt[0] * nt[1] + y * nt[0] + x
    assert len(seen) == len(threads)  # DispatchThreadID 无重复

    # 2) 硬件限制校验(cs_5_0:≤1024/组,Z≤64;cs_4_x:≤768,Z≤1)
    ok, _ = validate_numthreads((1024, 1, 1), CS5)
    assert ok
    ok, _ = validate_numthreads((32, 32, 1), CS5)   # 1024 恰好
    assert ok
    ok, _ = validate_numthreads((33, 33, 1), CS5)   # 1089 超限
    assert not ok
    ok, _ = validate_numthreads((64, 1, 65), CS5)    # Z=65 超限
    assert not ok
    ok, _ = validate_numthreads((768, 1, 1), CS4X)
    assert ok
    ok, _ = validate_numthreads((256, 1, 2), CS4X)   # cs_4_x 的 Z 只能是 1
    assert not ok
    ok, _ = validate_dispatch((65535, 1, 1), CS5)
    assert ok
    ok, _ = validate_dispatch((65536, 1, 1), CS5)
    assert not ok

    # 3) 树形归约:groupshared + barrier 求和正确(含随机数据)
    rng = random.Random(2026)
    for nt_i in ((16, 16, 1), (256, 1, 1), (8, 4, 2)):
        n = nt_i[0] * nt_i[1] * nt_i[2]
        vals = [rng.randint(-100, 100) for _ in range(n)]
        sim = ThreadGroupSim(nt_i)
        got = sim.run_reduction(vals)
        assert got == sum(vals), (nt_i, got, sum(vals))
        assert len(sim.shared) == n

    # 4) 屏障次序:线程 0 写 groupshared,屏障后其他线程才保证可见;
    #    无屏障时,消费线程若在线程 0 的写之前执行,读到旧值 0
    class SharedMem:
        def __init__(self):
            self.x = 0

        def with_barrier(self):
            # 阶段 1:线程 0 写;阶段 2(屏障后):线程 1 读 → 必见新值
            self.x = 42
            self.barrier_count += 1   # GroupMemoryBarrierWithGroupSync
            return self.x

        def without_barrier(self):
            # 交错:线程 1 的读被调度在线程 0 的写之前 → 读到旧值
            old = self.x              # 线程 1 先读
            self.x = 42               # 线程 0 后写
            return old

        barrier_count = 0

    sm = SharedMem()
    assert sm.with_barrier() == 42
    assert sm.barrier_count == 1
    sm2 = SharedMem()
    assert sm2.without_barrier() == 0  # 读发生在写之前 → 旧值

    # 5) 原子指令:非原子丢更新,InterlockedAdd 不丢
    vals = [10, 20, 30, 40]
    assert interlocked_add_example(vals, use_atomic=True) == 100
    assert interlocked_add_example(vals, use_atomic=False) == 40  # 只剩最后一个写
    assert CS4X["atomics"] is False and CS5["atomics"] is True

    print("ALL TESTS PASSED")
    print("Dispatch(5,3,2)+numthreads(10,8,3): GroupThreadID(7,5,0) -> "
          "DispatchThreadID(27,13,0), GroupIndex 57")
    print("reduction checks passed for (16,16,1) / (256,1,1) / (8,4,2)")


if __name__ == "__main__":
    main()

File: 01-游戏开发/02-渲染/图形管线/计算着色器执行模型/go/main.go
// 计算着色器执行模型:Go 复刻线程标识、硬件限制与树形归约。
// 依据 Microsoft Learn "Compute Shader Overview"/"numthreads"/"Dispatch":
// SV_GroupID/SV_GroupThreadID/SV_DispatchThreadID/SV_GroupIndex 四值自洽;
// cs_5_0 每组 ≤1024 线程、Z≤64、Dispatch 每维 ≤65535。
package main

import (
	"fmt"
	"math/rand"
)

type profile struct {
	maxThreadsPerGroup int
	maxZ               int
	maxDispatchDim     int
	atomics            bool
}

var cs5 = profile{1024, 64, 65535, true}
var cs4x = profile{768, 1, 65535, false}

// threadID:一个线程的四个系统值标识。
type threadID struct {
	groupID          [3]int // SV_GroupID
	groupThreadID    [3]int // SV_GroupThreadID
	dispatchThreadID [3]int // SV_DispatchThreadID
	groupIndex       int    // SV_GroupIndex
}

func newThreadID(numthreads, groupID, groupThreadID [3]int) threadID {
	dt := [3]int{}
	for i := 0; i < 3; i++ {
		dt[i] = groupID[i]*numthreads[i] + groupThreadID[i]
	}
	x, y := numthreads[0], numthreads[1]
	gi := groupThreadID[2]*x*y + groupThreadID[1]*x + groupThreadID[0]
	return threadID{groupID, groupThreadID, dt, gi}
}

func validateNumthreads(nt [3]int, p profile) bool {
	return nt[0]*nt[1]*nt[2] <= p.maxThreadsPerGroup && nt[2] <= p.maxZ
}

func validateDispatch(d [3]int, p profile) bool {
	for _, v := range d {
		if v > p.maxDispatchDim {
			return false
		}
	}
	return true
}

// groupReduce:groupshared + 屏障的树形归约(分阶段模拟,阶段内顺序无关)。
func groupReduce(numthreads [3]int, values []int) int {
	n := numthreads[0] * numthreads[1] * numthreads[2]
	if len(values) != n {
		panic("values length mismatch")
	}
	shared := make([]int, n)
	copy(shared, values) // 阶段 0:每线程写自己的值
	// GroupMemoryBarrierWithGroupSync(阶段边界)
	for stride := n / 2; stride > 0; stride /= 2 {
		for i := 0; i < stride; i++ {
			shared[i] += shared[i+stride]
		}
		// 屏障:下一级归约必须看到本级的全部写入
	}
	return shared[0]
}

func check(label string, cond bool, detail string) {
	if !cond {
		panic("FAIL " + label + ": " + detail)
	}
	fmt.Println("ok -", label)
}

func main() {
	// 1) MS Learn 文档标准例子:Dispatch(5,3,2)+numthreads(10,8,3)
	nt := [3]int{10, 8, 3}
	tid := newThreadID(nt, [3]int{2, 1, 0}, [3]int{7, 5, 0})
	check("DispatchThreadID", tid.dispatchThreadID == [3]int{27, 13, 0},
		fmt.Sprintf("%v", tid.dispatchThreadID))
	check("GroupIndex 57", tid.groupIndex == 57, fmt.Sprintf("%d", tid.groupIndex))

	// 全量枚举:DispatchThreadID 唯一,GroupIndex 与组内坐标一一对应
	total := 0
	seen := map[[3]int]bool{}
	for gz := 0; gz < 2; gz++ {
		for gy := 0; gy < 3; gy++ {
			for gx := 0; gx < 5; gx++ {
				for tz := 0; tz < nt[2]; tz++ {
					for ty := 0; ty < nt[1]; ty++ {
						for tx := 0; tx < nt[0]; tx++ {
							t := newThreadID(nt, [3]int{gx, gy, gz}, [3]int{tx, ty, tz})
							seen[t.dispatchThreadID] = true
							want := tz*nt[0]*nt[1] + ty*nt[0] + tx
							if t.groupIndex != want {
								panic(fmt.Sprintf("group index mismatch: %d vs %d", t.groupIndex, want))
							}
							total++
						}
					}
				}
			}
		}
	}
	check("total threads", total == 5*3*2*10*8*3, fmt.Sprintf("%d", total))
	check("dispatch IDs unique", len(seen) == total, fmt.Sprintf("%d", len(seen)))

	// 2) 硬件限制(cs_5_0 / cs_4_x)
	check("cs5 1024 ok", validateNumthreads([3]int{32, 32, 1}, cs5), "")
	check("cs5 1089 fail", !validateNumthreads([3]int{33, 33, 1}, cs5), "")
	check("cs5 z65 fail", !validateNumthreads([3]int{64, 1, 65}, cs5), "")
	check("cs4x 768 ok", validateNumthreads([3]int{768, 1, 1}, cs4x), "")
	check("cs4x z2 fail", !validateNumthreads([3]int{256, 1, 2}, cs4x), "")
	check("dispatch 65535 ok", validateDispatch([3]int{65535, 1, 1}, cs5), "")
	check("dispatch 65536 fail", !validateDispatch([3]int{65536, 1, 1}, cs5), "")

	// 3) 树形归约正确性(随机数据)
	rng := rand.New(rand.NewSource(2026))
	for _, ntI := range [][3]int{{16, 16, 1}, {256, 1, 1}, {8, 4, 2}} {
		n := ntI[0] * ntI[1] * ntI[2]
		vals := make([]int, n)
		want := 0
		for i := range vals {
			vals[i] = rng.Intn(201) - 100
			want += vals[i]
		}
		got := groupReduce(ntI, vals)
		check(fmt.Sprintf("reduce %v", ntI), got == want,
			fmt.Sprintf("got=%d want=%d", got, want))
	}

	// 4) 原子性:非原子丢更新 vs InterlockedAdd
	vals := []int{10, 20, 30, 40}
	// 原子:逐个读改写,不丢
	atomicSum := 0
	for _, v := range vals {
		atomicSum += v
	}
	// 非原子:全部线程先读旧值 0,再各自写回 → 只剩最后写的
	nonAtomic := 0
	for _, v := range vals {
		nonAtomic = 0 + v // 读到的 counter 恒为 0
	}
	check("interlocked add", atomicSum == 100, fmt.Sprintf("%d", atomicSum))
	check("lost update", nonAtomic == 40, fmt.Sprintf("%d", nonAtomic))
	check("cs4x no atomics", !cs4x.atomics && cs5.atomics, "")

	fmt.Println("ALL TESTS PASSED")
}
